modules/rostests/apitests/acpi/Bus_PDO_EvalMethod.c

```diff
@@ -0,0 +1,495 @@
+                sizeof(ACPI_EVAL_OUTPUT_BUFFER) },
+};
+
+/* FUNCTIONS ******************************************************************/
+
+static
+NTSTATUS
+DrvCallAcpiDriver(
+    _In_ PVOID InputBuffer,
+    _In_ ULONG InputBufferLength,
+    _Out_opt_ PACPI_EVAL_OUTPUT_BUFFER OutputBuffer,
+    _In_ ULONG OutputBufferLength)
+{
+    PDO_DEVICE_DATA DeviceData;
+    IRP Irp;
+
+    DeviceData.AcpiHandle = NULL;
+
+    Irp.AssociatedIrp.SystemBuffer = InputBuffer;
+    Irp.MyStack.Parameters.DeviceIoControl.InputBufferLength = InputBufferLength;
+    Irp.MyStack.Parameters.DeviceIoControl.OutputBufferLength = OutputBufferLength;
+
+    return Bus_PDO_EvalMethod(&DeviceData, &Irp);
+}
+
+static
+VOID
+DrvEvaluateStmObject(
+    _In_ const EVAL_TEST_ENTRY* TestEntry,
+    _In_ PIDE_ACPI_TIMING_MODE_BLOCK TimingMode,
+    _In_ PUCHAR IdBlock,
+    _In_ PACPI_EVAL_INPUT_BUFFER_COMPLEX InputBuffer)
+{
+    PACPI_METHOD_ARGUMENT Argument, Argument2, Argument3;
+    NTSTATUS Status;
+    ULONG InputBufferSize, IrpBufferSize;
+
+    InputBufferSize = FIELD_OFFSET(ACPI_EVAL_INPUT_BUFFER_COMPLEX, Argument) +
+                      ACPI_METHOD_ARGUMENT_LENGTH(sizeof(*TimingMode)) +
+                      ACPI_METHOD_ARGUMENT_LENGTH(STM_ID_BLOCK_SIZE) +
+                      ACPI_METHOD_ARGUMENT_LENGTH(STM_ID_BLOCK_SIZE);
+
+    if (TestEntry->Flags & STM_TEST_FLAG_INVALID_SIZE_1)
+    {
+        InputBufferSize -= ACPI_METHOD_ARGUMENT_LENGTH(STM_ID_BLOCK_SIZE) +
+                           ACPI_METHOD_ARGUMENT_LENGTH(STM_ID_BLOCK_SIZE);
+    }
+
+    InputBuffer->MethodNameAsUlong = IDE_ACPI_SET_TIMING_MODE;
+    InputBuffer->Signature = ACPI_EVAL_INPUT_BUFFER_COMPLEX_SIGNATURE;
+    if (TestEntry->Flags & STM_TEST_FLAG_SUB_IN_BUFFER)
+    {
+        InputBuffer->Size = InputBufferSize - TestEntry->Value;
+    }
+    else if (TestEntry->Flags & STM_TEST_FLAG_SET_IN_BUFFER)
+    {
+        InputBuffer->Size = TestEntry->Value;
+    }
+    else
+    {
+        InputBuffer->Size = InputBufferSize;
+    }
+
+    if (TestEntry->Flags & STM_TEST_FLAG_CHANGE_ARG_COUNT)
+    {
+        InputBuffer->ArgumentCount = TestEntry->Value;
+    }
+    else
+    {
+        InputBuffer->ArgumentCount = 3;
+    }
+
+    /* Argument 1: The channel timing information block */
+    Argument = InputBuffer->Argument;
+    ACPI_METHOD_SET_ARGUMENT_BUFFER(Argument, TimingMode, sizeof(*TimingMode));
+
+    /* Argument 2: The ATA drive ID block */
+    Argument2 = ACPI_METHOD_NEXT_ARGUMENT(Argument);
+    ACPI_METHOD_SET_ARGUMENT_BUFFER(Argument2, IdBlock, STM_ID_BLOCK_SIZE);
+
+    /* Argument 3: The ATA drive ID block */
+    Argument3 = ACPI_METHOD_NEXT_ARGUMENT(Argument2);
+    ACPI_METHOD_SET_ARGUMENT_BUFFER(Argument3, IdBlock, STM_ID_BLOCK_SIZE);
+
+    if (TestEntry->Flags & STM_TEST_FLAG_INVALID_SIZE_2)
+    {
+        Argument2->DataLength = 0x500;
+        Argument3->DataLength = 0x500;
```

Comment:
Are these values arbitrary? or is there a `#define` or a `sizeof()` for them?

Reply:
Added more descriptive names.